File: LineParser.java
import java.util.*;

public class LineParser {
    private final List<Token> line;
    private final Token eol;
    private int cursor;
    private int writtenLineNumber;

    public LineParser(List<Token> line) {
        this.line = appendEol(line);
        this.eol = this.line.get(this.line.size() - 1);
        this.cursor = 0;
    }

    public Command parse() {
        Token first = getCurrentToken();
        if (first.getKind() != Token.Kind.INTEGER_LITERAL) {
            syntaxError("line must start with a line number");
        }
        writtenLineNumber = Integer.parseInt(first.getText());
        acceptIt();
        return parseCommand();
    }

    private Command parseCommand() {
        if (getCurrentToken().getKind() == Token.Kind.EOL) {
            return new Command.Empty(writtenLineNumber);
        }
        switch (getCurrentToken().getText()) {
            case "let": return parseLet();
            case "print": return parsePrint();
            case "input": return parseInput();
            case "for": return parseFor();
            case "next": return parseNext();
            case "if": return parseIf();
            case "goto": return parseGoto();
            case "gosub": return parseGoSub();
            case "return": return parseReturn();
            case "end": return parseEnd();
            default: syntaxError(
                "expected a keyword (let, print, ...) but saw " +
                    getCurrentToken().getText()
            );
        }
        return null; // shouldn't be reachable, but javac insists
    }

    private Command.Let parseLet() {
        acceptIt();
        Token identifier = accept(Token.Kind.IDENTIFIER);
        accept(Token.Kind.EQ);
        Expression exp = parseExpression();
        accept(Token.Kind.EOL);
        return new Command.Let(writtenLineNumber, identifier.getText(), exp);
    }

    private Command.Print parsePrint() {
        acceptIt();
        List<Object> values = new ArrayList<>();
        while (getCurrentToken().getKind() != Token.Kind.EOL) {
            switch (getCurrentToken().getKind()) {
                case STRING_LITERAL:
                case COMMA:
                case SEMICOLON:
                    values.add(acceptIt());
                    break;
                default:
                    values.add(parseExpression());
                    break;
            }
        }
        accept(Token.Kind.EOL);
        return new Command.Print(writtenLineNumber, values);
    }

    private Command.Input parseInput() {
        acceptIt();
        String prompt = null;
        if (getCurrentToken().getKind() == Token.Kind.STRING_LITERAL) {
            prompt = acceptIt().getText();
        }
        List<String> variableNames = new ArrayList<>();
        variableNames.add(accept(Token.Kind.IDENTIFIER).getText());
        while (getCurrentToken().getKind() == Token.Kind.COMMA) {
            acceptIt();
            variableNames.add(accept(Token.Kind.IDENTIFIER).getText());
        }
        accept(Token.Kind.EOL);
        return new Command.Input(writtenLineNumber, prompt, variableNames);
    }

    private Command.For parseFor() {
        acceptIt();
        Token identifier = accept(Token.Kind.IDENTIFIER);
        accept(Token.Kind.EQ);
        Expression from = parseExpression();
        accept(Token.Kind.TO);
        Expression to = parseExpression();
        accept(Token.Kind.EOL);
        return new Command.For(writtenLineNumber, identifier.getText(), from, to);
    }

    private Command.Next parseNext() {
        acceptIt();
        Token identifier = accept(Token.Kind.IDENTIFIER);
        accept(Token.Kind.EOL);
        return new Command.Next(writtenLineNumber, identifier.getText());
    }

    private Command.If parseIf() {
        acceptIt();
        BooleanExpression exp = parseBooleanExpression();
        accept(Token.Kind.THEN);
        Command cmd = parseCommand();
        return new Command.If(writtenLineNumber, exp, cmd);
    }

    private Command.Goto parseGoto() {
        acceptIt();
        Token target = accept(Token.Kind.INTEGER_LITERAL);
        accept(Token.Kind.EOL);
        return new Command.Goto(writtenLineNumber,
            Integer.parseInt(target.getText()));
    }

    private Command.GoSub parseGoSub() {
        acceptIt();
        Token target = accept(Token.Kind.INTEGER_LITERAL);
        accept(Token.Kind.EOL);
        return new Command.GoSub(writtenLineNumber,
            Integer.parseInt(target.getText()));
    }

    private Command.Return parseReturn() {
        acceptIt();
        accept(Token.Kind.EOL);
        return new Command.Return(writtenLineNumber);
    }

    private Command.End parseEnd() {
        acceptIt();
        accept(Token.Kind.EOL);
        return new Command.End(writtenLineNumber);
    }

    private BooleanExpression parseBooleanExpression() {
        Expression left = parseExpression();
        Token t = getCurrentToken();
        Operator op = null;
        if (t.getKind() == Token.Kind.LT) {
            op = Operator.LT;
        } else if (t.getKind() == Token.Kind.EQ) {
            op = Operator.EQ;
        } else if (t.getKind() == Token.Kind.GT) {
            op = Operator.GT;
        } else {
            syntaxError("expected '<', '=' or '>', but saw " + t.getText());
        }
        acceptIt();
        Expression right = parseExpression();
        return new BooleanExpression(left, op, right);
    }

    private Expression parseExpression() {
        Expression term1 = parseTerm();
        while (getCurrentToken().getKind() == Token.Kind.PLUS ||
            getCurrentToken().getKind() == Token.Kind.MINUS) {
            Operator op = null;
            if (getCurrentToken().getKind() == Token.Kind.PLUS) {
                op = Operator.PLUS;
            } else {
                op = Operator.MINUS;
            }
            acceptIt();
            Expression term2 = parseTerm();
            term1 = new Expression.Binary(term1, op, term2);
        }
        return term1;
    }

    private Expression parseTerm() {
        Expression factor1 = parseFactor();
        while (getCurrentToken().getKind() == Token.Kind.TIMES ||
            getCurrentToken().getKind() == Token.Kind.DIV) {
            Operator op = null;
            if (getCurrentToken().getKind() == Token.Kind.TIMES) {
                op = Operator.TIMES;
            } else {
                op = Operator.DIV;
            }
            acceptIt();
            Expression factor2 = parseFactor();
            factor1 = new Expression.Binary(factor1, op, factor2);
        }
        return factor1;
    }

    private Expression parseFactor() {
        if (getCurrentToken().getKind() == Token.Kind.IDENTIFIER) {
            String variableName = acceptIt().getText();
            return new Expression.Identifier(variableName);
        } else if (getCurrentToken().getKind() == Token.Kind.INTEGER_LITERAL) {
            int value = Integer.parseInt(acceptIt().getText());
            return new Expression.Literal(value);
        } else if (getCurrentToken().getKind() == Token.Kind.L_PAREN) {
            acceptIt();
            Expression exp = parseExpression();
            accept(Token.Kind.R_PAREN);
            return exp;
        } else if (getCurrentToken().getKind() == Token.Kind.MINUS) {
            acceptIt();
            Expression exp = parseFactor();
            return new Expression.Binary(new Expression.Literal(0), Operator.MINUS, exp);
        } else {
            syntaxError("expected variable, integer, '(' or '-', but saw " +
                getCurrentToken().getText());
        }
        return null; // shouldn't be reachable, but javac insists
    }

    private Token getCurrentToken() {
        if (cursor > line.size()) {
            return eol;
        }
        return line.get(cursor);
    }

    private Token acceptIt() {
        Token t = getCurrentToken();
        cursor++;
        return t;
    }

    private Token accept(Token.Kind kind) {
        Token t = getCurrentToken();
        if (t.getKind() != kind) {
            syntaxError(String.format("expected %s but saw %s",
                kind, t.getKind()));
        }
        cursor++;
        return t;
    }

    private List<Token> appendEol(List<Token> line) {
        List<Token> prepped = new ArrayList<>(line);
        if (prepped.isEmpty()) {
            prepped.add(new Token(Token.Kind.EOL, "<eol>", 0, 0));
        } else {
            Token last = prepped.get(prepped.size() - 1);
            if (last.getKind() != Token.Kind.EOL) {
                int col = last.getColumnNumber() + last.getText().length();
                prepped.add(new Token(Token.Kind.EOL, "<eol>", last.getLineNumber(), col));
            }
        }
        return prepped;
    }

    private void syntaxError(String msg) {
        Token t = getCurrentToken();
        String fullMsg = String.format("[%d:%d] syntax error: %s",
            t.getLineNumber(), t.getColumnNumber(), msg);
        throw new RuntimeException(fullMsg);
    }
}

File: Interpreter/LineParser.cs
using System;
using System.Collections.Generic;

namespace Basic
{
    public class LineParser
    {
        private readonly IList<Token> _line;
        private readonly Token _eol;
        private int _cursor;
        private int _writtenLineNumber;

        public LineParser(IList<Token> line)
        {
            _line = AppendEol(line);
            _eol = _line[^1];
            _cursor = 0;
        }

        public Command Parse()
        {
            if (CurrentToken.Type == TokenType.Eol)
            {
                return new Command.Empty(-1); // Empty line gets phony line number
            }

            bool startsWithInt = (CurrentToken.Type == TokenType.IntegerLiteral)
                && int.TryParse(CurrentToken.Text, out _writtenLineNumber);
            if (!startsWithInt)
            {
                throw new SyntaxErrorException(
                    "line must start with an integer line number",
                    CurrentToken.Line,
                    CurrentToken.Column
                );
            }

            AcceptIt();
            Command cmd = ParseCommand();
            Accept(TokenType.Eol);
            return cmd;
        }

        private Command ParseCommand()
        {
            if (CurrentToken.Type == TokenType.Eol)
            {
                return new Command.Empty(_writtenLineNumber);
            }

            switch (CurrentToken.Type)
            {
                case TokenType.Let: return ParseLet();
                case TokenType.Print: return ParsePrint();
                case TokenType.Input: return ParseInput();
                case TokenType.For: return ParseFor();
                case TokenType.Next: return ParseNext();
                case TokenType.If: return ParseIf();
                case TokenType.Goto: return ParseGoto();
                case TokenType.Gosub: return ParseGosub();
                case TokenType.Return: return ParseReturn();
                case TokenType.End: return ParseEnd();
                default: break;
            }

            throw new SyntaxErrorException(
                $"expected a keyword (let, print, ...) but saw '{CurrentToken.Text}'",
                CurrentToken.Line,
                CurrentToken.Column
            );
        }

        private Command ParseLet()
        {
            AcceptIt(); // Consume "let"
            Token ident = Accept(TokenType.Identifier);
            Accept(TokenType.Eq);
            Expression exp = ParseExpression();
            Accept(TokenType.Eol);
            return new Command.Let(_writtenLineNumber, ident.Text, exp);
        }

        private Command ParsePrint()
        {
            AcceptIt(); // Consume "print"
            List<Expression> values = new List<Expression>();
            while (CurrentToken.Type != TokenType.Eol)
            {
                switch (CurrentToken.Type)
                {
                    case TokenType.Comma:
                        values.Add(new Expression.StringLiteral("\t"));
                        AcceptIt();
                        break;
                    case TokenType.Semicolon:
                        AcceptIt();
                        break;
                    default:
                        values.Add(ParseExpression());
                        break;
                }
            }
            return new Command.Print(_writtenLineNumber, values);
        }

        private Command ParseInput()
        {
            AcceptIt(); // Consume "input"

            string? prompt = null;
            if (CurrentToken.Type == TokenType.StringLiteral)
            {
                prompt = CurrentToken.Text;
                AcceptIt();
            }

            List<string> variableNames = new List<string>();
            variableNames.Add(Accept(TokenType.Identifier).Text);
            while (CurrentToken.Type == TokenType.Comma)
            {
                AcceptIt();
                variableNames.Add(Accept(TokenType.Identifier).Text);
            }

            return new Command.Input(_writtenLineNumber, prompt, variableNames);
        }

        private Command ParseFor()
        {
            AcceptIt(); // Consume "for"
            string variableName = Accept(TokenType.Identifier).Text;
            Accept(TokenType.Eq);
            Expression from = ParseExpression();
            Accept(TokenType.To);
            Expression to = ParseExpression();
            return new Command.For(_writtenLineNumber, variableName, from, to);
        }

        private Command ParseNext()
        {
            AcceptIt(); // Consume "next"
            Token ident = Accept(TokenType.Identifier);
            return new Command.Next(_writtenLineNumber, ident.Text);
        }

        private Command ParseIf()
        {
            AcceptIt(); // Consume "if"
            Expression test = ParseExpression();
            Accept(TokenType.Then);
            Command then = ParseCommand();
            return new Command.If(_writtenLineNumber, test, then);
        }

        private Command ParseGoto()
        {
            AcceptIt(); // Consume "goto"
            Token targetLine = Accept(TokenType.IntegerLiteral);
            return new Command.Goto(_writtenLineNumber, int.Parse(targetLine.Text));
        }

        private Command ParseGosub()
        {
            AcceptIt(); // Consume "gosub"
            Token targetLine = Accept(TokenType.IntegerLiteral);
            return new Command.Gosub(_writtenLineNumber, int.Parse(targetLine.Text));
        }

        private Command ParseReturn()
        {
            AcceptIt(); // Consume "return"
            return new Command.Return(_writtenLineNumber);
        }

        private Command ParseEnd()
        {
            AcceptIt(); // Consume "end"
            return new Command.End(_writtenLineNumber);
        }

        private Expression ParseExpression()
        {
            Expression left = ParseArithmeticExpression();

            if (IsRelationalOp(CurrentToken))
            {
                Operator op = GetOperator(AcceptIt());
                Expression right = ParseArithmeticExpression();
                return new Expression.Binary(left, op, right);
            }
            return left;
        }

        private Expression ParseArithmeticExpression()
        {
            Expression term1 = ParseTerm();
            while (CurrentToken.Type == TokenType.Plus ||
                   CurrentToken.Type == TokenType.Minus)
            {
                Operator op = GetOperator(AcceptIt());
                Expression term2 = ParseTerm();
                term1 = new Expression.Binary(term1, op, term2);
            }
            return term1;
        }

        private Expression ParseTerm()
        {
            Expression factor1 = ParseFactor();
            while (CurrentToken.Type == TokenType.Mult ||
                   CurrentToken.Type == TokenType.Div ||
                   CurrentToken.Type == TokenType.Mod)
            {
                Operator op = GetOperator(AcceptIt());
                Expression factor2 = ParseFactor();
                factor1 = new Expression.Binary(factor1, op, factor2);
            }
            return factor1;
        }

        private Expression ParseFactor()
        {
            switch (CurrentToken.Type)
            {
                case TokenType.Identifier:
                    string variableName = CurrentToken.Text;
                    AcceptIt();
                    return new Expression.Identifier(variableName);
                case TokenType.IntegerLiteral:
                    int value = int.Parse(CurrentToken.Text);
                    AcceptIt();
                    return new Expression.IntegerLiteral(value);
                case TokenType.StringLiteral:
                    string text = CurrentToken.Text;
                    AcceptIt();
                    return new Expression.StringLiteral(text);
                case TokenType.ParenL:
                    AcceptIt();
                    Expression exp = ParseExpression();
                    Accept(TokenType.ParenR);
                    return exp;
                case TokenType.Minus:
                    AcceptIt();
                    Expression factor = ParseFactor();
                    return new Expression.Binary(new Expression.IntegerLiteral(0), Operator.Minus, factor);
                default: break;
            }

            throw new SyntaxErrorException(
                $"expected variable, literal, '(' or '-', but saw '{CurrentToken.Text}'",
                CurrentToken.Line,
                CurrentToken.Column
            );
        }

        private Operator GetOperator(Token token)
        {
            switch (token.Type) {
                case TokenType.Lt: return Operator.Lt;
                case TokenType.Eq: return Operator.Eq;
                case TokenType.Gt: return Operator.Gt;
                case TokenType.Plus: return Operator.Plus;
                case TokenType.Minus: return Operator.Minus;
                case TokenType.Mult: return Operator.Mult;
                case TokenType.Div: return Operator.Div;
                case TokenType.Mod: return Operator.Mod;
                default: throw new NotImplementedException("operator: " + token.Text);
            }
        }

        private Token CurrentToken => (_cursor < _line.Count) ? _line[_cursor] : _eol;

        private static bool IsRelationalOp(Token token)
        {
            return token.Type == TokenType.Eq ||
               token.Type == TokenType.Gt ||
               token.Type == TokenType.Lt;
        }

        private static bool IsArithmeticOp(Token token)
        {
            return token.Type == TokenType.Plus || 
               token.Type == TokenType.Minus ||
               token.Type == TokenType.Mult ||
               token.Type == TokenType.Div ||
               token.Type == TokenType.Mod;
        }

        private static IList<Token> AppendEol(IList<Token> tokens)
        {
            IList<Token> prepped = new List<Token>(tokens);
            if (prepped.Count == 0)
            {
                prepped.Add(new Token(TokenType.Eol, "<eol>", 1, 0));
            }
            else
            {
                Token last = tokens[^1];
                if (last.Type != TokenType.Eol)
                {
                    int col = last.Column + last.Text.Length;
                    prepped.Add(new Token(TokenType.Eol, "<eol>", last.Line, col));
                }
            }
            return prepped;
        }

        private Token AcceptIt()
        {
            Token t = CurrentToken;
            _cursor++;
            return t;
        }

        private Token Accept(TokenType expectedType)
        {
            Token t = CurrentToken;
            if (CurrentToken.Type != expectedType)
            {
                throw new SyntaxErrorException(
                    $"expected {expectedType} but saw {CurrentToken.Type}",
                    CurrentToken.Line,
                    CurrentToken.Column
                );
            }
            _cursor++;
            return t;
        }
    }
}